### ISSUE TITLE
Multinode dashboard extension

### ISSUE THREAD

**Issue by ArtemAlagizov** [utkuozdemir/nvidia_gpu_exporter#27]
Hi, great project, mate, thanks!

I don't seem to be able to get data over multiple nodes, though.
In our company we have an on-premise cluster of PCs with GPUs, on top of which there is a k8s cluster and we are looking for a proper monitoring solution.
I use 3 nodes in my test setup -> 1 master and 2 nodes. All of them are to pick up GPU workloads.

i used Helm chart to deploy the exporters
```bash
helm install ozdemir utkuozdemir/nvidia-gpu-exporter -f nvidia-gpu-utku-ozdemir-values.yml
```

with the following values to allow deploying to master nodes as well
```
tolerations:
  - key: node-role.kubernetes.io/master
    effect: NoSchedule
    operator: Exists

service:
  type: NodePort
  nodePort: 30699
```
_NOTE: By the way, as it turns out it's not possible to set a nodePort in Helm values, are you planning on adding this feature by any chance? would be very convenient._

exporters seem to be deployed successfully
```bash
vvcServiceAccount@k8s-master-node0:~$ k get pod -o wide
NAME                                                      READY   STATUS    RESTARTS   AGE   IP          NODE               NOMINATED NODE   READINESS GATES
utku-ozdemir-nvidia-gpu-exporter-6gt5r                         1/1     Running   1          19d   10.44.0.1   k8s-worker-node1   <none>           <none>
utku-ozdemir-nvidia-gpu-exporter-7jqgs                         1/1     Running   1          19d   10.32.0.2   k8s-master-node0   <none>           <none>
utku-ozdemir-nvidia-gpu-exporter-m9h72                         1/1     Running   0          19d   10.36.0.1   k8s-worker-node2   <none>           <none>
```

logs from all 3 exporters are the following
```
level=info ts=2021-08-13T15:08:03.434Z caller=main.go:65 msg="Listening on address" address=:9835
ts=2021-08-13T15:08:03.435Z caller=log.go:124 level=info msg="TLS is disabled." http2=false
```

prometheus specs for scraping 
```bash
prometheus:
  prometheusSpec:
    additionalScrapeConfigs:
    - job_name: nvidia_gpu_exporter
      static_configs:
        - targets: [ '10.0.10.3:31585' ] 
```

with 31585 being the random nodePort assigned to the sevice and 10.0.10.3 being the ip address of the master node
```bash
NAME                               TYPE       CLUSTER-IP       EXTERNAL-IP   PORT(S)          AGE
utku-ozdemir-nvidia-gpu-exporter   NodePort   10.106.182.203   <none>        9835:31585/TCP   19d
```

i used this grafana dashboard to display the collected data: https://grafana.com/grafana/dashboards/14574

the results are good:
![image](https://user-images.githubusercontent.com/15367507/131820275-7399be5a-5bad-413f-b7ed-cb14ad6045ee.png)

but there is only one node in the dropdown (one of the worker nodes).
am i missing something in my setup or is it not possible to have all nodes in the dropdown list to have an overview per node?

ideally, it would be awesome to also have a summary dashboard with averaged metrics to monitor the whole cluster at once.
do you by any chance plan on developing something like that or maybe know some dashboards that already do that?
i tried the official one (nvidia/gpu-operator with https://grafana.com/grafana/dashboards/12239 as a dashboard) but it is not nearly as impressive as yours and also it had a bunch of empty charts and "no data" type of situations.

so, to sum up
* would be cool to add nodePort option to Helm chart values
* is it possible to have an overview of multiple nodes?

thanks!

**Comment by ArtemAlagizov** [comment]
i can surely provide any extra logs or setup details if needed

**Comment by utkuozdemir** [comment]
Hi, glad that you liked the project, and thanks for the detailed description!

I didn't have a chance to test it on a multi-node cluster with GPUs so I was expecting to face this kind of issues.

I'll add NodePort support to the chart, no problem.

In the meantime: Is `10.0.10.3` in your scrapeConfig target the IP of the `Service`? If so, it might be the problem. Prometheus has to scrape **each and every DaemonSet po**d every N seconds (whatever period you have configured) to get all the metrics from all GPUs. When it only scrapes a service every N seconds, it will hit one of the pods randomly (maybe round-robin, maybe with some other LB algorithm).

I guess you don't use Prometheus Operator, since I see that you don't use the `ServiceMonitor` custom resource in the chart. In any case, it wouldn't work correctly either - it would cause the same problem you are facing.

For the moment, you can try to add each pod IP of the DS as the target to prometheus, and then check the metrics/Grafana. That way we can verify that the issue is actually this one I described. After that, we can look into ways to solve it (Pod IPs is not a real solution because they change all the time).

**Comment by utkuozdemir** [comment]
Seems I have to add this into the chart instead of ServiceMonitor: https://stackoverflow.com/a/66482645/1005102

If you cannot use Prometheus Operator, NodePorts can be useful for you since nodes have stable IPs that can be used for scraping.

**Comment by ArtemAlagizov** [comment]
thanks for the quick response and the information!

i actually do use prometheus-operator, i just excluded not relevant info in my opinion, that's the full prometheus section of the values file for prometheus-operator:
```bash
prometheus:
  service:
    type: NodePort
    nodePort: 30658
  serviceMonitorSelectorNilUsesHelmValues: false
  prometheusSpec:
    additionalScrapeConfigs:
    - job_name: nvidia_gpu_exporter
      static_configs:
        - targets: [ '10.0.10.3:31585' ]
```

i will try adding each pod IP as a target for prometheus and will get back to you soon with the results.

**Comment by ArtemAlagizov** [comment]
10.0.10.3 is the ip address of the PC where master node is deployed by the way

**Comment by ArtemAlagizov** [comment]
you identified the problem correctly indeed!
after adding each pod IP as target for prometheus, i can see all 3 nodes in the dropdown.

![image](https://user-images.githubusercontent.com/15367507/131839244-15636673-10ae-48c1-99ed-46fd8f89f3af.png)

the config i used
```bash
prometheus:
  service:
    type: NodePort
    nodePort: 30658
  serviceMonitorSelectorNilUsesHelmValues: false
  prometheusSpec:
    additionalScrapeConfigs:
    - job_name: nvidia_gpu_exporter
      static_configs:
        - targets: [ '10.44.0.1:9835', '10.32.0.2:9835', '10.36.0.1:9835' ]
```

**Comment by utkuozdemir** [comment]
I released chart version `0.3.0` with support for hostPort and nodePort options. Should be visible here in ~30mins https://artifacthub.io/packages/helm/utkuozdemir/nvidia-gpu-exporter

(About ServiceMonitor - turns out there was no problem with that - its behavior is to scrape all matching pods already)

In your use-case I recommend using a `hostPort` and putting `<NODE_IP>:<HOST_PORT>` for each node to your static scrape config targets.

Using `nodePort` will not work properly because it will end as the same behavior - Service will still load-balance to the pods.

**Comment by ArtemAlagizov** [comment]
thanks very much, @utkuozdemir, works perfectly!
this issue if fully covered, awesome support.

that's the settings i used:
```bash
# ozdemir-nvidia-gpu-exporter Helm values.yml

hostPort:
  enabled: true
  port: 31585
```

```bash
# prometheus operator values

prometheus:
  ...
  prometheusSpec:
    additionalScrapeConfigs:
    - job_name: nvidia_gpu_exporter
      static_configs:
        - targets: [ '10.0.10.3:31585', '10.0.10.7:31585', '10.0.10.8:31585' ]
```
where 10.0.10.3, 10.0.10.7, 10.0.10.8 are ip addresses of my nodes.

**Comment by utkuozdemir** [comment]
Glad that I could help, thanks for reporting and all the information.
